### PR TITLE
refactor: broker client doesn't start topology gossip server

### PR DESCRIPTION
**Pull request by lenaschoenburg**
`BrokerClient` is supposed to be an isolated component that can be started multiple times. This fixes an issue where the `BrokerClient` would start a `ClusterTopologyGossiper` (indirectly, by starting `GatewayClusterTopologyService` in order to connect it to a new `BrokerTopologyManager`). This is not allowed because the `ClusterTopologyGossiper` can only run once.
To fix this, the `GatewayClusterTopologyService`, `ClusterTopologyManagerService` and `BrokerTopologyManager` are now managed by Spring and a `BrokerClient` is created with a given `BrokerTopologyManager` instead of building a new one for every client.

### CODE REVIEW

**Comment by npepinpe** [comment]
can we just rewrite the broker client plsssssss :smile:

File: dist/src/main/java/io/camunda/zeebe/gateway/TopologyServices.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Autowired
+  public TopologyServices(final ActorScheduler scheduler, final AtomixCluster atomixCluster) {
+    this.scheduler = scheduler;
+    clusterMembershipService = atomixCluster.getMembershipService();
+    clusterCommunicationService = atomixCluster.getCommunicationService();
+  }
+
+  @Bean
+  GatewayClusterTopologyService gatewayClusterTopologyService() {
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/gateway/TopologyServices.java:37]: Didn't this start it also on the broker? If yes, then there will still be two gossiper running in broker, Right? One from this `GatewayClusterTopologyService` and one from the `ClusterTopologyManagerService`.

File: dist/src/main/java/io/camunda/zeebe/broker/TopologyServices.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    clusterMembershipService = atomixCluster.getMembershipService();
+  }
+
+  @Bean
+  BrokerTopologyManager brokerTopologyManager() {
+    final var brokerTopologyManager =
+        new BrokerTopologyManagerImpl(clusterMembershipService::getMembers);
+    scheduler.submitActor(brokerTopologyManager).join();
+    clusterMembershipService.addListener(brokerTopologyManager);
+    // TODO: register with gossip
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/broker/TopologyServices.java:37]: :question: This is registered later in the broker, right? So this TODO is not valid?

File: dist/src/main/java/io/camunda/zeebe/gateway/TopologyServices.java (hunk 1 above)
Comment (deepthidevaki) [reply dist/src/main/java/io/camunda/zeebe/gateway/TopologyServices.java:37]: Got it. There is only started for standalone gateway because of spring only scan this package when starting StandaloneGateway. :+1:

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java [hunk 3]
```diff
@@ -213,22 +198,15 @@ private void updateMetrics(final BrokerClusterState topology) {
...
-          LOG.debug(
-              "Received new cluster topology with clusterSize {}", newTopology.getClusterSize());
-          topology.set(newTopology);
+  public void onTopologyUpdated(final ClusterTopology clusterTopology) {
+    if (clusterTopology.isUninitialized()) {
+      return;
+    }
+    LOG.debug("Received new cluster topology with clusterSize {}", clusterTopology.clusterSize());
+    updateTopology(
+        topology -> {
```
Comment (deepthidevaki) [inline gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java:207]: :upside_down_face: My IDE shows warning `Rename "topology" which hides the field declared at line 32.`

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java [hunk 4]
```diff
@@ -17,59 +17,44 @@
...
 public final class BrokerTopologyManagerImpl extends Actor
-    implements BrokerTopologyManager,
-        ClusterMembershipEventListener,
-        GatewayClusterTopologyService.Listener {
+    implements BrokerTopologyManager, ClusterMembershipEventListener, TopologyUpdateListener {
 
   private static final Logger LOG = Loggers.GATEWAY_LOGGER;
-
-  private final AtomicReference<BrokerClusterStateImpl> topology;
+  private volatile BrokerClusterStateImpl topology = new BrokerClusterStateImpl();
```
Comment (deepthidevaki) [inline gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java:32]: :question: Why was this changed from `AtomicReference`? Agree it doesnot make much difference as concurrent threads only read them. But I think the semantics is clear with AtomicReference.

File: dist/src/main/java/io/camunda/zeebe/broker/BrokerClientConfiguration.java [hunk 5]
```diff
@@ -21,23 +23,30 @@ final class BrokerClientConfiguration {
...
-        cluster.getCommunicationService(),
-        scheduler);
+    final var brokerClient =
+        new BrokerClientImpl(
+            config.getGateway().getCluster().getRequestTimeout(),
+            cluster.getMessagingService(),
+            cluster.getEventService(),
+            scheduler,
+            topologyManager);
+    brokerClient.start().forEach(ActorFuture::join);
```
Comment (deepthidevaki) [inline dist/src/main/java/io/camunda/zeebe/broker/BrokerClientConfiguration.java:49]: :question: I guess it is ok to start it here. Any idea why it was not before?

**Review by deepthidevaki — APPROVED** [review]
:rocket: Nice approach to inject them as spring components.

File: dist/src/main/java/io/camunda/zeebe/broker/TopologyServices.java (hunk 2 above)
Comment (oleschoenburg) [reply dist/src/main/java/io/camunda/zeebe/broker/TopologyServices.java:37]: Oh yeah, that's a leftover :+1:

File: gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java (hunk 4 above)
Comment (oleschoenburg) [reply gateway/src/main/java/io/camunda/zeebe/gateway/impl/broker/cluster/BrokerTopologyManagerImpl.java:32]: IMO the previous approach with `AtomicReference` was not sound anyway because we usually just did a `get`, then made some calculations and then `set` the result. So this already depended on single-threaded execution through the actor to be safe.

**Comment by oleschoenburg** [comment]
bors merge
